### DISCUSSION TITLE
Using DNS for quick distribution of filter list updates

### DISCUSSION THREAD

**Discussion by sekrause** [uBlockOrigin/uBlock-issues#2873] [Ideas]
**The problem:** You have a large website that updates it's anti-adblock measures multiple times a day. You want to quickly distribute changes to the filter list to your users, but the 12 hours expiry of the "quick fixes" list is not low enough. But you can't lower the expiry of the list because that would overload the CDN.

**The idea:** Maybe it's possible to use DNS to distribute quick changes to filter lists and get free caching from DNS resolvers.

There is now a widespread, global infrastructure of DNS-over-HTTPS resolvers. Some ([for example Cloudflare](https://developers.cloudflare.com/1.1.1.1/encryption/dns-over-https/make-api-requests/dns-json/)) can even provide JSON reponses for easy querying from within JavaScript. DNS [TXT records](https://en.wikipedia.org/wiki/TXT_record) allow the distribution of arbitrary text and DNS resolvers cache the record for the duration of the TTL.

A filter list could have a special syntax which defines DNS records that contain additional filters, something like:

```
||thatsite.com^$dnsfilters=thatsite.ublockdns.example.com
```

Now each time I visit `thatsite.com` uBlock Origin would fetch the TXT records from `thatsite.ublockdns.example.com` (using a DNS-over-HTTPS server configured by the user) and use the filters from the TXT records as well. The DNS records would have a reasonably low TTL and each DNS resolver would cache the records for that duration (with the cache being shared among all of the resolver's users), resulting in a much lower load on the upstream servers.

Since DNS records can't be very large they would only contain the most recent changes until every user will have them from the main filter list anyway. So if the "quick fixes" list has an expiration of 12 hours hours a filter can be rotated out of the DNS records after 12 hours.

**Comment by uBlock-user** [comment]
Not possible in Chromium, maybe looked at Firefox if possible this way.

**Comment by stephenhawk8054** [comment]
@gorhill @gwarser Since I'm not familiar with DNS stuff, I told OP to move the discussion to github.

**Comment by ghost** [comment]
uBO will need to have own domain and use subdomains for this, like `thatsite.com.ubo-home.test`?

**Reply by stephenhawk8054** [reply]
Is it possible with cloudflare pages?

**Reply by ghost** [reply]
`thatsite.com` is not important, can be a hash of domain or predefined IDs. Problem is with `ubo-home.test`, the **home** part.

**Reply by sekrause** [reply]
> uBO will need to have own domain and use subdomains for this, like `thatsite.com.ubo-home.test`?

Yes, it would require some new infrastructure:

* A domain name that you have under your control (ability to add subdomains).
* A nameserver that hosts the DNS records.

**Reply by Yuki2718** [reply]
I doubt @gorhill agrees. Once you get domain, you owe responsibility of it. Whenever we abandon the domain, anyone can abuse that. Being free from such responsibility has been central to this project.

**Reply by sekrause** [reply]
> Once you get domain, you owe responsibility of it. Whenever we abandon the domain, anyone can abuse that.

I understand. By only referencing the domain in the filter list you could easily revoke or change the use of this domain with a simple update of the filter list. Or even only implement such a feature and let somebody else have the responsibility of providing such a domain (it doesn't need to be done by the uBO team, just like there are filter lists from other teams).

**Comment by DandelionSprout** [comment]
Normally I'd suggest giving a tip about it to https://github.com/AdguardTeam/AdGuardHome/issues, but development on that tool has borderline collapsed since around February, with issue reports being left to linger there for years.

However, distributing list updates by DNS sound like nonsense to me. I at least don't think DNS works that way. I also fully expect there to be some sort of max limit to how long a TXT-type query ping result can be; very likely not long enough to handle potentially 4-digit amounts of entries at a time.

uBO supporting hours for update intervals was new to me (They only supported 1 day or longer for around 10 years), so since they're trying out new things anyway to the best of my current knowledge (which may not be perfect), why not set the interval for Quick Fixes down from 12 to 6 hours during the YouTube war thingie?

**Reply by sekrause** [reply]
> I also fully expect there to be some sort of max limit to how long a TXT-type query ping result can be; very likely not long enough to handle potentially 4-digit amounts of entries at a time.

Yes, the limits are quite low, usually a maximum of 255 characters per string or 4k for the whole TXT record. That's why my suggestion was to use different subdomains per filter list domain and only temporarily add the newest changes to DNS. For example after [this commit](https://github.com/uBlockOrigin/uAssets/commit/886064b7c2014b6345225e3213b585bd7908a553) (where only one line changes) you would have a single 139 character long TXT record (or just 121 characters if you strip out the domain part) for 12 hours.

I agree that distributing whole filters lists through DNS would be nonsense.

**Reply by stephenhawk8054** [reply]
> set the interval for Quick Fixes down from 12 to 6 hours during the YouTube war thingie?

The number of requests are too high currently, cloudflare and jsdelivr have 1.1B and 730M total requests last month respectively, not counting github pages and statically.io, so at least 3B requests last month and the quick fix list counts the most.

**Reply by stephenhawk8054** [reply]
> maximum of 255 characters

Hmmm... 2 yt filters already cost ~240 characters

```adb
www.youtube.com##+js(trusted-replace-fetch-response,/"adPlacements.*?AD_PLACEMENT_KIND_START.*?"\}\}\}\]\,/, ,url:player?key=)
www.youtube.com##+js(trusted-replace-fetch-response, /"adSlots.*?SLOT_TYPE_PLAYER_BYTES.*?\}\]\}\}\]\,/, , url:player?key=)
```

**Reply by ghost** [reply]
> why not set the interval for Quick Fixes down from 12 to 6 hours during the YouTube war thingie?

Does not matter when CDNs update content on the servers every two days.

**Reply by stephenhawk8054** [reply]
@gwarser I think cloudflare is quite fast? It contains today's fix already: https://ublockorigin.pages.dev/filters/quick-fixes.txt

**Reply by ajayyy** [reply]
Maybe the TXT record could contain the last modified time if the quick fixes list

**Reply by sekrause** [reply]
> Hmmm... 2 yt filters already cost ~240 characters

That would be two TXT string values though (with each having a 255 character limit). It would only become a bit difficult if a single filter entry would be longer than 255 entries, [then the usual way is to split it into two values](https://support.google.com/a/answer/11613097).

> Maybe the TXT record could contain the last modified time if the quick fixes list

That could make the load on the CDN worse because now you would have all uBO clients updating the filter list at the same time. It only really helps if you get rid of the frequent HTTP refreshes.

**Comment by WurstCommander** [comment]
Is a peer-to-peer solution for acquiring the ublock filterfiles compliant with Chrome's guidelines? Given the high number of uBlock installations, sharing the filter list among clients could be achieved effortlessly.

I mean, there are bittorrent/webtorrent extensions. No need for a domain or a centralized solution with peer-2-peer. Just an idea.
